ddio/file.cpp

```diff
@@ -0,0 +1,155 @@
+/*
+ * Descent 3
+ * Copyright (C) 2024 Descent Developers
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <array>
+#include <filesystem>
+#include <fstream>
+#include <iterator>
+
+#include "IOOps.h"
+#include "ddio.h"
+#include "pserror.h"
+
+struct LockFileContent {
+  std::array<char, 4> tag = {'L', 'O', 'C', 'K'};
+  int32_t pid = -1;
+};
+
+std::ostream &operator<<(std::ostream &output, const LockFileContent &header) {
+  std::copy(header.tag.begin(), header.tag.end(), std::ostream_iterator<char>(output));
+  D3::bin_write(output, header.pid);
+  return output;
+}
+
+std::istream &operator>>(std::istream &input, LockFileContent &header) {
+  input.read(reinterpret_cast<char*>(header.tag.data()), 4);
+  D3::bin_read(input, header.pid);
+  return input;
+}
+
+int ddio_CheckLockFile(const std::filesystem::path& lock_filename, int pid) {
+  if(!std::filesystem::is_directory(lock_filename.parent_path())) {
+    return -1;
+  }
+
+  if (!std::filesystem::exists(lock_filename)) {
+    return 1;
+  }
+
+  LockFileContent lock_content;
+  try {
+    std::ifstream lockfile(lock_filename, std::ios::binary);
+    lockfile >> lock_content;
+    lockfile.close();
+  } catch (std::exception &e) {
+    return -2;
+  }
+
+  if (lock_content.tag != std::array{'L', 'O', 'C', 'K'}) {
```

Comment:
the "LOCK" array should be a const array (defined in an anonymous namespace?) so you only have to declare it once